### ISSUE TITLE
Optimization not progressing with PyTorch

### ISSUE THREAD

**Issue by ruihao-li** [AgnostiqHQ/covalent#645]
### Environment

- **Covalent version**: 0.110.1
- **Python version**: 3.8.8
- **Operating system**: MacOS 12.4

### What is happening?

This issue is similar to Issue #638: when using PyTorch with Covalent, the gradients are not evaluated properly. Note that this is not unique to the Dask executor, but the same thing happens with the local executor as well. This issue is also seen in the [MNIST classification tutorial](https://covalent.readthedocs.io/en/latest/tutorials/machine_learning/mnist_classifier.html), where the loss stays more or less constant over multiple training epochs.  

### How can we reproduce the issue?

```
@ct.electron
def cost(a):
    l = torch.sum(a ** 2)
    return l

@ct.electron
def move_forward(optimizer,point):
    loss = cost(point)
    optimizer.zero_grad()
    loss.backward()
    optimizer.step()
    return loss, optimizer

@ct.lattice(executor=dask_executor)
def optimize(point,steps):
    optimizer = torch.optim.Adam([point], lr=.2)
    for i in range(steps):
        loss, optimizer = move_forward(optimizer,point)
    return point

init_point = torch.randn(3, requires_grad=True)
print(init_point)
dispatch_id = ct.dispatch(optimize)(init_point, 10)
result = ct.get_result(dispatch_id=dispatch_id, wait=True)
```
One would see that the result is the same as the init_point.

### What should happen?

Result should be different from init_point if gradient descent is performed properly.

**Comment by cjao** [comment]
See #646 for MNIST. I would restructure your code so that Optimizer creation and training happen in the same electron:
```

@ct.electron
def cost(a):
    l=torch.sum(a ** 2)
    return l

@ct.electron
def move_forward(optimizer,point):
    loss = cost(point)
    optimizer.zero_grad()
    loss.backward()
    optimizer.step()
    return loss,optimizer

@ct.electron
def train(point, steps):
    optimizer = torch.optim.Adam([point], lr=.2)
    l = []
    for i in range(steps):
        loss,optimizer=move_forward(optimizer,point)
        l.append(loss)
    return l

@ct.electron
def move_forward_does_not_works(point,optimizer):
    loss = cost(point)
    optimizer.zero_grad()
    loss.backward()
    optimizer.step()
    return loss,optimizer

@ct.lattice
def optimize(point,steps):
    return train(point=point, steps=steps)
x=torch.tensor([1,1.],requires_grad=True)
r=ct.dispatch(optimize)(x,10)
print(ct.get_result(r,wait=True).result)
```
I get
```
[tensor(2., requires_grad=True), tensor(1.2800, requires_grad=True), tensor(0.7257, requires_grad=True), tensor(0.3358, requires_grad=True), tensor(0.1020, requires_grad=True), tensor(0.0062, requires_grad=True), tensor(0.0180, requires_grad=True), tensor(0.0964, requires_grad=True), tensor(0.1971, requires_grad=True), tensor(0.2826, requires_grad=True)]
```

**Comment by cjao** [comment]
By the way, the electron decorator on `cost` does nothing; electrons within another electron are executed as ordinary python functions.

**Comment by ruihao-li** [comment]
Thanks for looking into this and the detailed explanations in [#646](https://github.com/AgnostiqHQ/covalent/issues/646) @cjao! I'll close this issue now.
